packages/rsocket-core/src/RSocketMachine.js

```diff
@@ -330,8 +331,116 @@ class RSocketMachineImpl<D, M> implements RSocketMachine<D, M> {
   }
 
   requestChannel(payloads: Flowable<Payload<D, M>>): Flowable<Payload<D, M>> {
```

Comment:
I'm torn on this and will think on it a bit. I like the simplicity of the API yet I totally agree that the implementation gets weird, waiting on the first payload and deferring its delivery. On the other hand, the caller would have to implement something similar assuming they're also dealing with stream sources, and I'd almost rather keep that complexity encapsulated. Like I said, I'm torn :)

Reply:
I'm less torn now - I'd much rather keep the simpler API (just the flowable<payload> input) and encapsulate the complexity. The Java version does this and it seems much better to write the operator once rather than inflict it on users. Plus, it makes it makes it much easier to compose requestChannel calls without having to the slicing every time. In a separate PR, I can carve out just the "switch on first" logic into a helper or something in the Flowables package.

Reply:
So the reason we changed if for rsocket-cpp was for API/user ergonomics. We always use the request payload for metadata/routing and not be part of the actual channel stream. I think we just our usages just differ here.

Agree to not change it now, we might also get away with supporting both APIs or something similar.